### PR TITLE
JumpSystem cleanup

### PR DESCRIPTION
**Pull request by isaacsas**
As a first step before working on making JumpSystems support coupled ODEs/SDEs, I wanted to clean up the code a bit. This 

1. Adds support for `collect_vars!` for `JumpSystems`.
2. Cleans up the constructor to be more in line with how the code works for other system types. If at some point refactoring of common code in constructors is done it should now be easier to see what is different in JumpSystems.

Note that JumpSystems would still need some more work to support hierarchical modeling (i.e. sub-systems and scoping).

### CODE REVIEW

**Comment by isaacsas** [comment]
My plan is to next make the `ArrayPartition` that stores the equations also have a component for equations, which will be used to generate an `ODESystem` during `ODEProblem(::JumpSystem, ...)` to be used in making a coupled ODE-jump system.

**Review by AayushSabharwal — COMMENTED** [review]
Looks good. Could you check if removing the `if !(sys isa JumpSystem)` [here](https://github.com/SciML/ModelingToolkit.jl/blob/76805bea9036d56a17d667c0325e444ceda707e2/src/systems/abstractsystem.jl#L926-L934) now works? It was added because `collect_vars!` didn't work

**Comment by isaacsas** [comment]
Will update later for that. I think we need a dispatch on the trait I added in `collect_scoped_vars!`.

**Comment by isaacsas** [comment]
Should be updated now.

**Comment by isaacsas** [comment]
Lots of failures here, but as far as I can tell none are related to this PR. I formatted with JuliaFormatter v1.0.62 so not sure why that is failing either. Locally it reports there is nothing to format when run on the repo.

File: src/systems/jumps/jumpsystem.jl [hunk 1]
```diff
@@ -179,30 +206,42 @@ function JumpSystem(eqs, iv, unknowns, ps;
...
+
     JumpSystem{typeof(ap)}(Threads.atomic_add!(SYSTEM_COUNT, UInt(1)),
-        ap, value(iv), unknowns, ps, var_to_name, observed, name, description, systems,
+        ap, iv′, us′, ps′, var_to_name, observed, name, description, systems,
         defaults, connector_type, disc_callbacks, parameter_dependencies,
         metadata, gui_metadata, checks = checks)
 end
 
+##### MTK dispatches for JumpSystems #####
+eqtype_supports_collect_vars(j::MassActionJump) = true
```
Comment (AayushSabharwal) [inline src/systems/jumps/jumpsystem.jl:221]: What is the point of this trait? It doesn't seem to be used anywhere.

Comment (isaacsas) [reply src/systems/jumps/jumpsystem.jl:221]: https://github.com/SciML/ModelingToolkit.jl/blob/76805bea9036d56a17d667c0325e444ceda707e2/src/utils.jl#L503

Comment (isaacsas) [reply src/systems/jumps/jumpsystem.jl:221]: It is already used to make collect_scoped_vars work with non-Equation equations that have the trait.

Comment (isaacsas) [reply src/systems/jumps/jumpsystem.jl:221]: More generally, by using a trait like this that by default opts systems/features out of supporting new functionality, it gives an easy way for downstream packages like Catalyst (or even other system types like JumpSystem) to subsequently turn the feature on, and helps avoid hard coded checks like `isa JumpSystem` or `isa Equation`.

**Comment by isaacsas** [comment]
@ChrisRackauckas I think this is good to merge now -- the test failures all seem unrelated.

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by ChrisRackauckas** [comment]
This looks good for now. Though I think the path forward really is that a symbolic JumpSystem should have its variables as discretes and use the hybrid system stuff when possible. The reason is the purely discrete variables would then not be in the ODE when the two are mixed, that could be a big performance win. We probably need to sync about this.

**Comment by isaacsas** [comment]
Happy to chat about that at some point. Maybe we can talk at the next office hours. 

That would require changes in JumpProcesses too presumably? I thought what we are setting up here — really in the other WIP PR — would be pretty similar to what one would normally write by hand.

**Pull request merged**